### PR TITLE
rafttest: fix TestRestart timeout

### PR DESCRIPTION
**Pull request by gyuho**
Fix https://github.com/coreos/etcd/issues/6290.
1. Send X number of proposals to leader
2. Before proposals are completely processed, stop leader
3. Restart old leader as follower
4. Some of the proposals are still being sent to old leader
5. If the old leader(now follower) doesn't recognize the
   new leader yet, the proposal channel is nil, so it will
   block forever in rafttest

### CODE REVIEW

**Comment by gyuho** [comment]
@xiang90 Please review.

This only fixes https://github.com/coreos/etcd/issues/6290.

We still have a few more test failures in rafttest.
I will address them in different PRs.

File: raft/rafttest/node.go [hunk 1]
```diff
@@ -57,16 +58,44 @@ func startNode(id uint64, peers []raft.Peer, iface iface) *node {
 	return n
 }
 
+func waitLeader(nodes []*node) int {
+	for {
+		lead := raft.None
+		leadIdx := -1
+		for i, nd := range nodes {
+			if i == 0 {
+				lead = nd.lead
```
Comment (xiang90) [inline raft/rafttest/node.go]: it seems unsafe to access lead without any locking.

File: raft/rafttest/node.go [hunk 2]
```diff
@@ -79,6 +108,10 @@ func (n *node) start() {
 				}
 				n.Advance()
 			case m := <-n.iface.recv():
+				// previous leader failure
+				if m.Type == raftpb.MsgProp && n.id == n.lead && lead == 0 {
```
Comment (xiang90) [inline raft/rafttest/node.go]: i do not quite understand this. n.id == n.lead -> n is the leader. what is lead==0?

File: raft/rafttest/node_test.go [hunk 3]
```diff
@@ -87,6 +87,43 @@ func TestRestart(t *testing.T) {
 	}
 }
 
+func TestStopLeader(t *testing.T) {
```
Comment (xiang90) [inline raft/rafttest/node_test.go]: can you add a comment about what does this test do?

File: raft/rafttest/node.go (hunk 2 above)
Comment (xiang90) [inline raft/rafttest/node.go]: we probably should just do

``` go
if lead == raft.None {
    // drop proposals when there is no leader
    continue
}
```

Comment (xiang90) [inline raft/rafttest/node.go]: also for proposals, we should do `go n.Step()` like what do for etcd.

File: raft/rafttest/node_test.go (hunk 3 above)
Comment (gyuho) [inline raft/rafttest/node_test.go]: Sure, will add.

**Comment by xiang90** [comment]
I do not think this fully fix the problem. There is still a chance for dropping proposals. That test will not block, but line 84 will fail since we commit less than 1206.

**Comment by gyuho** [comment]
Right, I will rework on this. Thanks!

**Comment by gyuho** [comment]
@xiang90 Right, I am still seeing proposals dropping, so less commit count.

Do you think we should tune the `rafttest` tests?

Still, those tests failures do not reveal anything wrong about our raft implementation, I think.

**Comment by xiang90** [comment]
Closing in favor of https://github.com/coreos/etcd/pull/6807

**Pull request closed without merging**